### PR TITLE
[Permissions] records result of added, removed and found permissions

### PR DESCRIPTION
**Pull request by thenamankumar**
closes: https://github.com/sourcegraph/sourcegraph/issues/47076

This PR records added, removed and found permissions count for each permission sync job in the database. 
## Test plan

- unit tests added for both perms syncer worker and db store. 
image

### CODE REVIEW

File: migrations/frontend/1675367314_add_results_to_permission_sync_jobs/up.sql [hunk 1]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE permission_sync_jobs ADD COLUMN IF NOT EXISTS permissions_added Integer;
+ALTER TABLE permission_sync_jobs ADD COLUMN IF NOT EXISTS permissions_removed Integer;
+ALTER TABLE permission_sync_jobs ADD COLUMN IF NOT EXISTS permissions_found Integer;
```
Comment (sashaostrikov) [inline migrations/frontend/1675367314_add_results_to_permission_sync_jobs/up.sql:3]: Mr. Nit Pick back in town:
```suggestion
ALTER TABLE permission_sync_jobs ADD COLUMN IF NOT EXISTS permissions_added integer;
ALTER TABLE permission_sync_jobs ADD COLUMN IF NOT EXISTS permissions_removed integer;
ALTER TABLE permission_sync_jobs ADD COLUMN IF NOT EXISTS permissions_found integer;
```

File: internal/database/permission_sync_jobs.go [hunk 2]
```diff
@@ -281,6 +282,26 @@ WHERE id = %s AND state = 'queued' AND cancel IS FALSE
 	return nil
 }
 
+type SetPermissionsResult struct {
+	Added   int
+	Removed int
+	Found   int
+}
+
+func (s *permissionSyncJobStore) SaveResult(ctx context.Context, id int, result *SetPermissionsResult) error {
```
Comment (sashaostrikov) [inline internal/database/permission_sync_jobs.go:291]: I would name it less abstract, like: `SavePermsSyncJobResult`

File: internal/database/permission_sync_jobs.go [hunk 3]
```diff
@@ -387,6 +408,10 @@ type PermissionSyncJob struct {
 	Priority         PermissionSyncJobPriority
 	NoPerms          bool
 	InvalidateCaches bool
+
+	PermissionsAdded   *int
```
Comment (sashaostrikov) [inline internal/database/permission_sync_jobs.go:412]: Do you think pointers are necessary here?
`NumResets` and `NumFailures` doesn't have pointers and I think these new properties shouldn't too. We can easily distinguish "nil pointer" behaviour here if the `state` is not `completed` -- then these values are just zero-values and sync didn't complete successfully.

File: internal/database/permission_sync_jobs_test.go [hunk 4]
```diff
@@ -353,6 +353,47 @@ func TestPermissionSyncJobs_CancelQueuedJob(t *testing.T) {
...
+	}
+
+	logger := logtest.Scoped(t)
+	db := NewDB(logger, dbtest.NewDB(logger, t))
+	ctx := context.Background()
+
+	store := PermissionSyncJobsWith(logger, db)
+	reposStore := ReposWith(logger, db)
+
+	// create repo
```
Comment (sashaostrikov) [inline internal/database/permission_sync_jobs_test.go:368]: nit: let's stick to "first capital letter and full stop in the end" comment style 😁

**Review by sashaostrikov — CHANGES_REQUESTED** [review]
LGTM!

I have a couple of concerns about tests:
- I can see some unnecessary changes in `[perms_syncer_worker_test.go](https://github.com/sourcegraph/sourcegraph/pull/47354/files#diff-e9cb6b0347867a66b0cd7cdc8d01dcdde92ba27a8f69e94feef3891651701750)`, let's revert most part of them and leave only `SetPermissionsResult` not adding new syncs and changing current behaviour of tests.
- I think we need to add some tests to `perms_store_test.go`, because we've added some behaviour to related file, but it is not checked in tests (return variable is just skipped everywhere)
- Unrelated to tests: name of function saving the stats/results, can be less abstract

File: enterprise/cmd/repo-updater/internal/authz/perms_syncer_worker_test.go [hunk 5]
```diff
@@ -87,36 +90,39 @@ func TestPermsSyncerWorker_RepoSyncJobs(t *testing.T) {
...
+
 	// Adding user perms sync job, which should not be processed by current worker!
 	err = syncJobsStore.CreateUserSyncJob(ctx, user2.ID,
 		database.PermissionSyncJobOpts{Reason: database.ReasonRepoNoPermissions, Priority: database.HighPriorityPermissionSync, TriggeredByUserID: user1.ID})
 	require.NoError(t, err)
 
 	// Wait for all jobs to be processed.
 	timeout := time.After(60 * time.Second)
-	remainingRounds := 3
+	remainingRounds := 4
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/cmd/repo-updater/internal/authz/perms_syncer_worker_test.go:125]: fixed

File: enterprise/cmd/repo-updater/internal/authz/perms_syncer_worker_test.go [hunk 6]
```diff
@@ -126,7 +132,7 @@ loop:
 		for _, job := range jobs {
 			// We don't check job with ID=3 because it is a user sync job which is not
 			// processed by current worker.
-			if job.ID != 3 && (job.State == "queued" || job.State == "processing") {
+			if job.ID == 3 && (job.State == "queued" || job.State == "processing") {
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/cmd/repo-updater/internal/authz/perms_syncer_worker_test.go]: fixed

File: enterprise/cmd/repo-updater/internal/authz/perms_syncer_worker_test.go [hunk 7]
```diff
@@ -165,17 +171,32 @@ loop:
 			require.Equal(t, jobID, job.RepositoryID)
 		}
 
-		// Check that failed job has the failure message.
+		// Check that user sync job wasn't picked up by repo sync worker.
```
[the remark this answers is not in the document]
Comment (thenamankumar) [reply enterprise/cmd/repo-updater/internal/authz/perms_syncer_worker_test.go]: fixed

File: migrations/frontend/1675367314_add_results_to_permission_sync_jobs/up.sql (hunk 1 above)
Comment (thenamankumar) [reply migrations/frontend/1675367314_add_results_to_permission_sync_jobs/up.sql]: updated

File: internal/database/permission_sync_jobs.go (hunk 2 above)
Comment (thenamankumar) [reply internal/database/permission_sync_jobs.go]: renamed

File: internal/database/permission_sync_jobs.go (hunk 3 above)
Comment (thenamankumar) [reply internal/database/permission_sync_jobs.go]: updated

File: internal/database/permission_sync_jobs_test.go (hunk 4 above)
Comment (thenamankumar) [reply internal/database/permission_sync_jobs_test.go]: done

**Review by sashaostrikov — APPROVED** [review]
:shipit:

**Review by pjlast — APPROVED** [review]
🚢 🤡

**Pull request merged**